### PR TITLE
feat(pointpainting fusion): add pointpainting

### PR DESCRIPTION
**Pull request by tzhong518**
## Description

- Add pointpainting as part of the image_projection_based_fusion package for the perception.
- It uses lidar_centerpoint as a shared library.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by yukkysaito** [comment]
@tzhong518 Can yot use rebase command to remove unrelated changings

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 1]
```diff
@@ -4,14 +4,111 @@ project(image_projection_based_fusion)
...
+else()
+  message("CUDNN is NOT Available")
+  set(CUDNN_AVAIL OFF)
+endif()
+
+# Download trained models
+find_program(GDOWN_AVAIL "gdown")
+if(NOT GDOWN_AVAIL)
+  message("gdown: command not found. External files could not be downloaded.")
+endif()
```
Comment (yukke42) [inline perception/image_projection_based_fusion/CMakeLists.txt:73]: not use gdown anymore. pls change the urls after the download links are created. 
see https://github.com/autowarefoundation/autoware.universe/pull/938

File: perception/image_projection_based_fusion/src/fusion_node.cpp [hunk 2]
```diff
@@ -124,7 +125,7 @@ FusionNode<Msg, ObjType>::FusionNode(
     std::placeholders::_7, std::placeholders::_8, std::placeholders::_9));
 
   // publisher
-  pub_ptr_ = this->create_publisher<Msg>("output", rclcpp::QoS{1});
+  pub_ptr_ = this->create_publisher<Msg>("~/debug/pointcloud_painted", rclcpp::QoS{1});
```
Comment (yukke42) [inline perception/image_projection_based_fusion/src/fusion_node.cpp:128]: pls don't change the topic name for other launchers.

**Comment by yukke42** [comment]
@tzhong518 Cloud you add the documentation like roi_cluster_fusion or roi_detected_object_fusion ?

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IMAGE_PROJECTION_BASED_FUSION__POINTPAINTING_FUSION__POINTPAINTING_TRT_HPP_
+#define IMAGE_PROJECTION_BASED_FUSION__POINTPAINTING_FUSION__POINTPAINTING_TRT_HPP_
+
+#include <image_projection_based_fusion/pointpainting_fusion/voxel_generator.hpp>
+#include <lidar_centerpoint/centerpoint_trt.hpp>
+
+namespace centerpoint
```
Comment (yukke42) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp:21]: change `centerpoint` to `image_projection_based_fusion`

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    const DensificationParam & densification_param, const CenterPointConfig & config);
+
+  ~PointPaintingTRT();
+
+protected:
+  bool preprocess(
+    const sensor_msgs::msg::PointCloud2 & input_pointcloud_msg,
+    const tf2_ros::Buffer & tf_buffer) override;
+
+  std::unique_ptr<image_projection_based_fusion::VoxelGenerator> vg_ptr_pp{nullptr};
```
Comment (yukke42) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp:39]: change `vg_ptr_pp` to `vg_ptr_pp_` like other protected values

File: perception/image_projection_based_fusion/CMakeLists.txt (hunk 1 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/CMakeLists.txt:73]: Please use the links below.
https://awf.ml.dev.web.auto/perception/models/pointpainting/pts_backbone_neck_head_pointpainting.onnx
https://awf.ml.dev.web.auto/perception/models/pointpainting/pts_voxel_encoder_pointpainting.onnx

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 5]
```diff
@@ -4,14 +4,111 @@ project(image_projection_based_fusion)
...
+  # https://drive.google.com/file/d/GFILE_ID/view
+  message(STATUS "Checking and downloading ${FILE_NAME}")
+  set(FILE_PATH ${DATA_PATH}/${FILE_NAME})
+  if(EXISTS ${FILE_PATH})
+    file(MD5 ${FILE_PATH} EXISTING_FILE_HASH)
+    if(${FILE_HASH} EQUAL ${EXISTING_FILE_HASH})
+      message(STATUS "File already exists.")
+    else()
+      message(STATUS "File hash changes. Downloading now ...")
+      execute_process(COMMAND gdown --quiet https://drive.google.com/uc?id=${GFILE_ID} -O ${FILE_PATH})
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/CMakeLists.txt:88]: pls change to T4 AWS.
@eratostennis can you support?

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2021 TIER IV, Inc.
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/pointpainting_trt.hpp:1]:
```suggestion
// Copyright 2022 TIER IV, Inc.
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+  sensor_msgs::PointCloud2Modifier pcd_modifier(painted_pointcloud_msg);
+  pcd_modifier.clear();
+  painted_pointcloud_msg.width = tmp.width;
+  pcd_modifier.setPointCloud2Fields(
+    7, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
+    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
+    sensor_msgs::msg::PointField::FLOAT32);
+  painted_pointcloud_msg.point_step = 28;
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: '28' is 7(num fields)x4(float32)?

```
  const int num_fields = 7;
   pcd_modifier.setPointCloud2Fields(
    7, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
    sensor_msgs::msg::PointField::FLOAT32);
  painted_pointcloud_msg.point_step = num_fields * sizeof(sensor_msgs::msg::PointField::FLOAT32);
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/pointpainting_trt.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 TIER IV, Inc.
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/pointpainting_trt.cpp:1]:
```suggestion
// Copyright 2022 TIER IV, Inc.
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/voxel_generator.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
+// Copyright 2021 TIER IV, Inc.
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/voxel_generator.cpp:1]:
```suggestion
// Copyright 2022 TIER IV, Inc.
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/voxel_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+std::size_t VoxelGenerator::pointsToVoxels(
+  std::vector<float> & voxels, std::vector<int> & coordinates,
+  std::vector<float> & num_points_per_voxel)
+{
+  // voxels (float): (max_num_voxels * max_num_points_per_voxel * point_feature_size)
+  // coordinates (int): (max_num_voxels * point_dim_size)
+  // num_points_per_voxel (float): (max_num_voxels)
+
+  const std::size_t grid_size = config_.grid_size_z_ * config_.grid_size_y_ * config_.grid_size_x_;
+  std::vector<int> coord_to_voxel_idx(grid_size, -1);
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/voxel_generator.cpp:30]: it is may be better to std::optional
```
  std::vector<std::optional<int>> coord_to_voxel_idx(grid_size);
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+      continue;
+    }
+    // project
+    Eigen::Vector4d projected_point =
+      camera_projection * Eigen::Vector4d(*iter_x, *iter_y, *iter_z, 1.0);
+    Eigen::Vector2d normalized_projected_point = Eigen::Vector2d(
+      projected_point.x() / projected_point.z(), projected_point.y() / projected_point.z());
+
+    // iterate 2d bbox
+    for (size_t i = 0; i < input_roi_msg.feature_objects.size(); ++i) {
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:174]:
```suggestion
    for (const auto & feature_object : input_roi_msg.feature_objects) {
```

File: perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp (hunk 7 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: 4 is the length of bytes of each field.

Comment (yukkysaito) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: I see.
How about this?
```
   constexpr int num_fields = 7;
   pcd_modifier.setPointCloud2Fields(
    num_fields, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
    sensor_msgs::msg::PointField::FLOAT32);
  painted_pointcloud_msg.point_step = num_fields * sizeof(float);
```

Comment (yukkysaito) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: I'd like to remove magic number like 7, 28

Comment (tzhong518) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: Got it. Thanks for the advice, I have pushed the changes.

Comment (yukkysaito) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:93]: image
Still left `7` :pray:

Comment (tzhong518) [reply perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp]: Sorry for the miss. I corrected it.

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/node.hpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    const sensor_msgs::msg::CameraInfo & camera_info,
+    sensor_msgs::msg::PointCloud2 & painted_pointcloud_msg) override;
+
+  void postprocess(sensor_msgs::msg::PointCloud2 & painted_pointcloud_msg) override;
+
+  rclcpp::Publisher<DetectedObjects>::SharedPtr obj_pub_ptr_;
+
+  float score_threshold_{0.0};
+  std::vector<std::string> class_names_;
+  std::vector<double> point_cloud_range;
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/node.hpp:53]:
```suggestion
  std::vector<double> pointcloud_range_;
```

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 13]
```diff
@@ -4,14 +4,115 @@ project(image_projection_based_fusion)
...
+      message(STATUS "Downloaded file hash: ${DOWNLOADED_FILE_HASH}")
+    endif()
+  else()
+    execute_process(COMMAND echo "not found ${FILE_NAME}")
+    message(STATUS "File doesn't exists. Downloading now ...")
+    execute_process(COMMAND curl -s "https://awf.ml.dev.web.auto/perception/models/pointpainting/pts_backbone_neck_head_pointpainting.onnx" -o ${FILE_PATH})
+    # file(MD5 ${FILE_PATH} DOWNLOADED_FILE_HASH)  # disable to pass ci
+    message(STATUS "Downloaded file hash: ${DOWNLOADED_FILE_HASH}")
+  endif()
+endfunction()
```
Comment (yukke42) [inline perception/image_projection_based_fusion/CMakeLists.txt:103]: pls be the same as [here](https://github.com/tier4/autoware.universe/blob/c8021d49ad2aa4dba26b3a81dd256fe196e10129/perception/lidar_centerpoint/CMakeLists.txt#L75-L105).

File: perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+{
+  const float score_threshold =
+    static_cast<float>(this->declare_parameter<double>("score_threshold", 0.4));
+  const float circle_nms_dist_threshold =
+    static_cast<float>(this->declare_parameter<double>("circle_nms_dist_threshold", 1.5));
+  // densification param
+  const std::string densification_world_frame_id =
+    this->declare_parameter("densification_world_frame_id", "map");
+  const int densification_num_past_frames =
+    this->declare_parameter("densification_num_past_frames", 1);
```
Comment (yukke42) [inline perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp:40]: densification aren't used yet. pls change the default parameter and set `0` to `densification_num_past_frames` if the value is more than 0 with a message.

File: perception/image_projection_based_fusion/docs/pointpainting-fusion.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+## Parameters
+
+### Core Parameters
+
+| Name                            | Type   | Default Value | Description                                                 |
+| ------------------------------- | ------ | ------------- | ----------------------------------------------------------- |
+| `score_threshold`               | float  | `0.4`         | detected objects with score less than threshold are ignored |
+| `densification_world_frame_id`  | string | `map`         | the world frame id to fuse multi-frame pointcloud           |
+| `densification_num_past_frames` | int    | `1`           | the number of past frames to fuse with the current frame    |
```
Comment (yukke42) [inline perception/image_projection_based_fusion/docs/pointpainting-fusion.md:42]: same as the default parameter.

**Comment by yukkysaito** [comment]
@tzhong518 can you resolve CI errors?

**Comment by tzhong518** [comment]
@yukkysaito Required CI test are passed.

**Comment by yukkysaito** [comment]
@tzhong518 Thank you :+1:

**Review by yukkysaito — APPROVED** [review]
LGTM
pls get approvement from @yukke42 @miursh

**Comment by yukke42** [comment]
I have checked to run pointpainting node, but the painted pointcloud needs to be subscribed to get the detected objects because of the following guard. 

https://github.com/autowarefoundation/autoware.universe/blob/458a32d0e132a01a092e3e3814176e564021e91c/perception/image_projection_based_fusion/src/fusion_node.cpp#L169-L171

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 16]
```diff
@@ -1,42 +1,179 @@
 cmake_minimum_required(VERSION 3.14)
 project(image_projection_based_fusion)
+add_compile_options(-Wall -Wno-unknown-pragmas)
```
Comment (yukke42) [inline perception/image_projection_based_fusion/CMakeLists.txt:3]:
```suggestion
add_compile_options(-Wno-unknown-pragmas)
```
It's in autoware_cmake.

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 17]
```diff
@@ -1,42 +1,179 @@
...
 
 find_package(autoware_cmake REQUIRED)
 autoware_package()
 
-find_package(OpenCV REQUIRED)
-find_package(Eigen3 REQUIRED)
```
Comment (yukke42) [inline perception/image_projection_based_fusion/CMakeLists.txt:8]: no delete

File: perception/image_projection_based_fusion/CMakeLists.txt [hunk 18]
```diff
@@ -1,42 +1,179 @@
...
+set(CUDA_VERBOSE ON)
```
Comment (yukke42) [inline perception/image_projection_based_fusion/CMakeLists.txt:8]:
```suggestion
set(CUDA_VERBOSE OFF)
```

File: perception/image_projection_based_fusion/docs/pointpainting-fusion.md [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+### Input
+
+| Name                  | Type                                                     | Description                                                                        |
+| --------------------- | -------------------------------------------------------- | ---------------------------------------------------------------------------------- |
+| `input`               | `sensor_msgs::msg::PointCloud2`                          | input pointcloud                                                                   |
+| `input/camera_infoID` | `sensor_msgs::msg::CameraInfo`                           | camera information to project 3d points onto image planes, `ID` is between 0 and 7 |
+| `input/roisID`        | `tier4_perception_msgs::msg::DetectedObjectsWithFeature` | ROIs from each image, `ID` is between 0 and 7                                      |
+| `input/image_rawID`   | `sensor_msgs::msg::Image`                                | images for visualization, `ID` is between 0 and 7                                  |
+
+| `
```
Comment (yukke42) [inline perception/image_projection_based_fusion/docs/pointpainting-fusion.md:24]: broken table format

File: perception/image_projection_based_fusion/docs/pointpainting-fusion.md [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+### Output
+
+| Name                 | Type                                                  | Description                                       |
+| -------------------- | ----------------------------------------------------- | ------------------------------------------------- | --- |
+| `output`             | `sensor_msgs::msg::PointCloud2`                       | painted pointclouda                               |
+| `~/output/objects`   | `autoware_auto_perception_msgs::msg::DetectedObjects` | detected objects                                  |     |
+| `output/image_rawID` | `sensor_msgs::msg::Image`                             | images for visualization, `ID` is between 0 and 7 |
```
Comment (yukke42) [inline perception/image_projection_based_fusion/docs/pointpainting-fusion.md:32]: broken table format

File: perception/image_projection_based_fusion/launch/pointpainting_fusion.launch.xml [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+      <param name="head_engine_path" value="$(var model_path)/pts_backbone_neck_head_$(var model_name).engine"/>
+      <param from="$(var model_param_path)"/>
+      <param name="rois_number" value="$(var input/rois_number)"/>
+
+      <!-- debug -->
+      <param name="debug_mode" value="$(var debug_mode)"/>
+    </node>
+  </group>
+  <!-- Rviz -->
+  <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen" args="-d $(find-pkg-share autoware_launch)/rviz/autoware.rviz -s $(find-pkg-share autoware_launch)/rviz/image/autoware.png"/>
```
Comment (yukke42) [inline perception/image_projection_based_fusion/launch/pointpainting_fusion.launch.xml:84]:
```suggestion
```
not launch rviz in this package

**Comment by yukke42** [comment]
please delete [this guard](https://github.com/tzhong518/autoware.universe/blob/feature/pointpainting_fusion/perception/image_projection_based_fusion/src/fusion_node.cpp#L170-L172)  and add to check subscription count [here](https://github.com/tzhong518/autoware.universe/blob/feature/pointpainting_fusion/perception/image_projection_based_fusion/src/fusion_node.cpp#L236).

File: perception/image_projection_based_fusion/config/pointpainting.param.yaml [hunk 22]
```diff
@@ -0,0 +1,10 @@
+/**:
+  ros__parameters:
+    class_names: ["CAR", "PEDESTRIAN", "BICYCLE"]
+    rename_car_to_truck_and_bus: true
+    point_feature_size: 6  # x, y, z and car, pedestrian, bicycle
+    max_voxel_size: 40000
+    point_cloud_range: [-76.8, -76.8, -3.0, 76.8, 76.8, 5.0]
+    voxel_size: [0.32, 0.32, 8.0]
+    downsample_factor: 2
```
Comment (yukke42) [inline perception/image_projection_based_fusion/config/pointpainting.param.yaml:9]:
```suggestion
    downsample_factor: 1
```

Cloud you confirm that the node runs successfully by this change?

**Comment by yukke42** [comment]
@tzhong518 Cloud you write where you fixed or change before you resolve conversation?

File: perception/image_projection_based_fusion/config/pointpainting.param.yaml (hunk 22 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/config/pointpainting.param.yaml:9]: yes, it runs on my machine.

File: perception/image_projection_based_fusion/launch/pointpainting_fusion.launch.xml (hunk 21 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/launch/pointpainting_fusion.launch.xml]: deleted.

File: perception/image_projection_based_fusion/CMakeLists.txt (hunk 18 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/CMakeLists.txt]: set to ON

File: perception/image_projection_based_fusion/CMakeLists.txt (hunk 17 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/CMakeLists.txt]: added it back.

File: perception/image_projection_based_fusion/CMakeLists.txt (hunk 16 above)
Comment (tzhong518) [reply perception/image_projection_based_fusion/CMakeLists.txt]: deleted.

**Review by yukke42 — APPROVED** [review]
LGTM

File: perception/image_projection_based_fusion/config/pointpainting.param.yaml (hunk 22 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/config/pointpainting.param.yaml]: fixed at [f267845](https://github.com/autowarefoundation/autoware.universe/pull/991/commits/f2678457b824ccba4514d0b548a3dce5baf7c717)

**Pull request merged**